### ISSUE TITLE
breakindent does not indent non-lists when breakindentopt=list:-1 is set

### ISSUE THREAD

**Issue by acc987** [vim/vim#11018]
### Steps to reproduce

1. Start gVim via `gvim -i NONE -u NONE -U NONE -N`.
2. Paste the following content into the current buffer and resize the window to make sure that the long line wraps:
```cpp
int my_function() {
	const int long_name = 0;
	return long_name + long_name + long_name + long_name + long_name + long_name + long_name + long_name;
}
```
![Wrapped contents](https://user-images.githubusercontent.com/39603492/187546173-24eda4b9-3401-40da-8793-c3e32ea6d251.png)
3. Issue `set breakindent`, the result looks as it should, which is as follows:
![SetBreakindent](https://user-images.githubusercontent.com/39603492/187546277-f80e5d34-ed8e-498d-8261-e2f762299a78.png)
4. Issue `set breakindentopt=list:-1`, this removes the indent even though it should not in my understanding:
![SetBreakindentopt](https://user-images.githubusercontent.com/39603492/187546464-c9652405-a848-4150-9412-5147dc9f96b2.png)

### Expected behaviour

Using `set breakindentopt=list:-1` should not break indentation for non-lists. (This issue does not seem to be exclusive to C++ code.)

### Version of Vim

9.0.0313

### Environment

openSUSE Leap 15.3

**Comment by acc987** [comment]
(Since I think this issue is mirrored to some mailing list: I have updated the original description to streamline the steps to reproduce.)

**Comment by lacygoill** [comment]
CC @habamax via https://github.com/vim/vim/pull/8594 and @chrisbra via https://github.com/vim/vim/pull/8564

**Comment by chrisbra** [comment]
The `list` suboption only works for lines that match your formatlistpat option. I don't think this is the case here.

**Comment by habamax** [comment]
So this is related to the lines 1020-1026 of indent.c:

```
    if (wp->w_briopt_list != 0)
    {
       if (wp->w_briopt_list > 0)
           bri += prev_list;
       else
           bri = prev_list;
    }
```

Where `bri = prev_list` is in question.

Basically, if briopt has `-1` we use indent of the `formatlistpat` which is not correct it seems.

So it should be changed in a way:

```
   if (wp->w_briopt_list != 0)
    {
           bri += prev_list;
    }
```

Plus `prev_list` should be calculated without initial indent for `list:-1`.

How would you do in `C` instead of:

```
prev_list = (*regmatch.endp - *regmatch.startp);
```

Something like (pseudo-code):
```
prev_list = len(strip_indent(*regmatch.value));
```

**Comment by habamax** [comment]
With the following, I was able to fix this:

```
diff --git a/src/indent.c b/src/indent.c
index d7a2e550a..b90bbacdb 100644
--- a/src/indent.c
+++ b/src/indent.c
@@ -998,7 +998,7 @@ get_breakindent_win(
 		    if (wp->w_briopt_list > 0)
 			prev_list = wp->w_briopt_list;
 		    else
-			prev_list = (*regmatch.endp - *regmatch.startp);
+			prev_indent = (*regmatch.endp - *regmatch.startp);
 		}
 		vim_regfree(regmatch.regprog);
 	    }
@@ -1017,13 +1017,8 @@ get_breakindent_win(
     bri += win_col_off2(wp);
 
     // add additional indent for numbered lists
-    if (wp->w_briopt_list != 0)
-    {
-	if (wp->w_briopt_list > 0)
+    if (wp->w_briopt_list != 0 && wp->w_briopt_list > 0)
 	    bri += prev_list;
-	else
-	    bri = prev_list;
-    }
 
     // indent minus the length of the showbreak string
     if (wp->w_briopt_sbr)
```


![image](https://user-images.githubusercontent.com/234774/187639286-cd93f0bf-12e1-4383-bc1c-a5015a3f10df.png)

**Comment by habamax** [comment]
One test with showbreak is failing though

**Comment by habamax** [comment]
Failing test seems like with an error, not sure how it was supposed to be.

It uses additional `briopt`, which is `shift:2` and expects it to be not shifted:

```
  call setline(1, ['* Congress shall make no law',
        \ '*** Congress shall make no law',
        \ '**** Congress shall make no law'])

...

  setl briopt=min:5,sbr,list:-1,shift:2
  setl showbreak=>
  redraw!
  let expect = [
	\ "* Congress shall    ",
	\ "> make no law       ",
	\ "*** Congress shall  ",
	\ ">   make no law     ",
	\ "**** Congress shall ",
	\ ">    make no law    ",
	\ ]


```

Removing `shift:2` makes it passed.

**Comment by habamax** [comment]
```
diff --git a/src/indent.c b/src/indent.c
index d7a2e550a..ec985ab3b 100644
--- a/src/indent.c
+++ b/src/indent.c
@@ -998,7 +998,7 @@ get_breakindent_win(
 		    if (wp->w_briopt_list > 0)
 			prev_list = wp->w_briopt_list;
 		    else
-			prev_list = (*regmatch.endp - *regmatch.startp);
+			prev_indent = (*regmatch.endp - *regmatch.startp);
 		}
 		vim_regfree(regmatch.regprog);
 	    }
@@ -1017,13 +1017,8 @@ get_breakindent_win(
     bri += win_col_off2(wp);
 
     // add additional indent for numbered lists
-    if (wp->w_briopt_list != 0)
-    {
-	if (wp->w_briopt_list > 0)
-	    bri += prev_list;
-	else
-	    bri = prev_list;
-    }
+    if (wp->w_briopt_list != 0 && wp->w_briopt_list > 0)
+	bri += prev_list;
 
     // indent minus the length of the showbreak string
     if (wp->w_briopt_sbr)
diff --git a/src/testdir/test_breakindent.vim b/src/testdir/test_breakindent.vim
index 7163f25bd..e65ac90b4 100644
--- a/src/testdir/test_breakindent.vim
+++ b/src/testdir/test_breakindent.vim
@@ -818,7 +818,7 @@ func Test_breakindent20_list()
 
   " check formatlistpat indent with different list level
   " showbreak and sbr
-  setl briopt=min:5,sbr,list:-1,shift:2
+  setl briopt=min:5,sbr,list:-1
   setl showbreak=>
   redraw!
   let expect = [

```

**Comment by habamax** [comment]
If it is OK, I can create a separate PR (but it would be the same as the patch provided earlier).

**Comment by acc987** [comment]
I can't comment on the fix itself as I am not familiar with the code, except maybe that `if (wp->w_briopt_list != 0 && wp->w_briopt_list > 0)` can be simplified, but I would suggest adding two more tests: One that is a copy of the one that would otherwise fail with your fix but this time actually tests that `setl briopt=min:5,sbr,list:-1,shift:2` does apply an additional shift of two to the right after the `'showbreak'` character. (I also think that the original test should not have passed in the first place.) And as for the second test, I would recommend that you also add a regression test, that demonstrates that non-list text is also indented when `breakindentopt` contains `list:-1`.

**Comment by habamax** [comment]
@acc987 for tests and condition it makes sense, indeed. Will do it.

**Comment by brammool** [comment]
Fixed with patch 9.0.0358
